bridge-stp.in

```diff
@@ -20,15 +20,70 @@ if [ $# -ne 2 ]; then
     exit 1
 fi
 bridge=$1
+
+# Define some relevant paths.
+mstpctl='@mstpctlfile@'
+mstpd='@mstpdfile@'
 pid_file='@mstpdpidfile@'
-conf_file='@bridgestpconffile@'
-bin_dir='@sbindir@'
-net_dir=/sys/class/net
+net_dir='/sys/class/net'
 
+# Set default config values.
+# If 'y', mstpd will be automatically started/stopped as needed.
+MANAGE_MSTPD='y'
+# Arguments to pass to mstpd when it is started.
+MSTPD_ARGS=''
 # A space-separated list of bridges for which MSTP should be used in place of
 # the kernel's STP implementation.  If empty, MSTP will be used for all bridges.
 MSTP_BRIDGES=''
 
+# Read the config.
+if [ -e '@bridgestpconffile@' ]; then
+    . '@bridgestpconffile@'
+fi
+
+# Ensure that mstpctl and mstpd exist and are executable.
+if [ -z "$mstpctl" ] || [ ! -x "$mstpctl" ]; then
+    echo "mstpctl binary does not exist or is not executable" >&2
+    exit 2
+fi
+if [ "$MANAGE_MSTPD" = 'y' ]; then
+    if [ -z "$mstpd" ] || [ ! -x "$mstpd" ]; then
+        echo "mstpd binary does not exist or is not executable" >&2
+        exit 2
+    fi
+fi
+
+# Determine whether mstpd should manage STP for the specified bridge.
+# Returns 0 if mstpd should manage STP for the specified bridge, or 1 if mstpd
+# should not manage STP for the specified bridge.
+is_mstp_bridge()
+{
+    if [ -z "$MSTP_BRIDGES" ]; then
+        return 0
+    fi
+    for b in $MSTP_BRIDGES; do
+        if [ "$1" = "$b" ]; then
+            return 0
+        fi
+    done
+    return 1
+}
+
+# Verify whether mstpctl can connect to mstpd.
+# Returns 1 if mstpctl cannot connect to mstpd.  Returns 0 if mstpctl is
+# successful or if it returns any other error.  Note that mstpctl may return an
+# error under normal conditions, for example if a bridge exists that has not
+# been added to mstpd.
+check_mstpctl()
+{
+    Out="$("$mstpctl" showbridge 2>&1)" && return 0
+    # if [ "$Out" contains "Couldn't connect to server" ]; then
+    if [ "$Out" != "${Out%Couldn\'t connect to server*}" ]; then
```

Comment:
this looks a bit brittle ; i'm not sure how well this handles [long-term] the check `Couldn't connect to server`
i admit I don't have a better idea atm ; will try to think of something